"Test connection" on LDAPS URI does not test TLS handshake

### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

admin/ui

### Describe the bug

When creating an LDAP federation in the web interface, there is a `Test Connection` button. However, this button only checks for a TCP handshake and tests neither the transport layer (TLS) nor the application layer. This can be seen when watching a packet capture on the backend LDAP server; whether using LDAP, LDAP with StartTLS, or LDAPS, keycloak only initiates a TCP handshake and then ends with a FIN ACK.

As a result, `Test Connection` will incorrectly report a functioning connection over URIs that would obviously fail at the transport level, such `ldaps://localhost:8080` (the keycloak HTTP interface).

Further, the `Test Authentication` button operates inconsistently. When a valid user is provided, it attempts to perform a handshake on TCP, security layer, and LDAP-- unless the settings are StartTLS with Truststore SPI set to Always, in which case it only performs a TCP handshake and disconnects.

### Version

22.0.5

### Expected behavior

1. `Test Connection` should validate TLS and StartTLS settings as part of the connection.
	- To confirm connectivity up to the application level, an anonymous LDAP bind should be attempted. Any LDAP response should be considered a success.
2. `Test Authentication` with bind type `none` should attempt an anonymous bind
3. `Test Authentication` should always attempt an LDAP bind and should always respect TLS / StartTLS settings

### Actual behavior

1. `Test Connection` only performs a TCP handshake and ignores TLS and startTLS settings
2. `Test Authentication` with bind type `none` only attempts a TCP connect, disregarding TLS settings
3. `Test Authentication` with bind type `simple`, StartTLS on and Truststore SPI set to Always only attempts a TCP connect and does not respect StartTLS settings or attempt a bind

### How to Reproduce?

In my testing, I did not set up trust for the LDAP TLS CA. This is not necessary as the incorrect behavior occurs before TLS verification would occur, and the lack of trust makes it easier to see in the Web UI what is actually happening.

| Environment | . |
| ---- | ---- |
| Keycloak Version | 22.05 (docker) |
| Keycloak Database | Postgres15 (docker) |
| LDAP Hostname | ldap.lab.local |
| Keycloak Hostname | keycloak.lab.local |
| Other | Keycloak does NOT trust LDAPS CA |

 1. Install and run a live packet capture on the LDAP server with filter 'host keycloak.domain.local'
 2. Verify normal packet patterns for ldap connection:

| Test | Command | Packet Capture |
| ---- | ---- | ---- |
| Plain LDAP | `ldapsearch -H ldap://ldap.lab.local:389 -s oneLevel "(cn=conn_test)"` | TCP handshake, LDAP response |
| LDAPS wrong port | `ldapsearch -H ldaps://ldap.lab.local:389 -s oneLevel "(cn=conn_test)"` | TCP Handshake, TLS handshake failure |
| LDAP StartTLS | `ldapsearch -Z -H ldap://ldap.lab.local:389 -s oneLevel "(cn=conn_test)"` | TCP Handshake, StartTLS issued |
| LDAPS correct port | `ldapsearch -H ldaps://ldap.lab.local:636 -s oneLevel "(cn=conn_test)"` | TCP Handshake, TLS handshake |

3. Create a new realm, and begin creating a new LDAP federation. Set Bind Type to 'none' and run both `Test Connection` and `Test Authentication`:

| Test               | URL | StartTLS                                                                  | Keycloak UI result    | Packet Capture                          |
| ------------------ | --- | ------------------------------------------------------------------------- | --- | ------------------------------- |
| Plain LDAP         | ldap://ldap.lab.local:389    | false                                                                          | Successfully connected to LDAP    | TCP handshake    |
| LDAPS wrong port   | ldaps://ldap.lab.local:389    | false   | Successfully connected to LDAP    | TCP handshake      |
| LDAP StartTLS      | ldap://ldap.lab.local:389    | True | Error when trying to connect to LDAP: 'UnknownError'    | TCP handshake |
| LDAPS correct port | ldaps://ldap.lab.local:636    | False   | Successfully connected to LDAP    | TCP handshake    |

4. Change Authentication type to 'simple' and provide a valid LDAP user. Test the `Test Authentication` button.

| Test | URL | StartTLS | Trust SPI | Keycloak UI result | Packet Capture |
| ---- | ---- | ---- | ---- | ---- | ---- |
| Plain LDAP | ldap://ldap.lab.local:389 | false | Any | Successfully connected to LDAP | TCP handshake, LDAP query |
| LDAPS wrong port | ldaps://ldap.lab.local:389 | false | Any | 'Socket Reset' | TCP handshake, TLS Handshake |
| LDAP StartTLS SPI Always | ldap://ldap.lab.local:389 | True | Always | 'UnknownError' | TCP handshake |
| LDAP StartTLS SPI Never | ldap://ldap.lab.local:389 | True | Never | 'SSL Handshake Failed' | TCP handshake, StartTLS |
| LDAPS correct port | ldaps://ldap.lab.local:636 | False | Any | 'SSL Handshake Failed' | TCP handshake, TLS handshake |


### Anything else?

The positioning of the `Test Connection` button is problematic because it's position below the StartTLS and TrustStore settings imply that it will test those settings as well, but it does not.

The fact that these buttons exist but perform inconsistently makes troubleshooting exceptionally difficult because there is no single way to validate only the lower layers and then validate only the authentication settings.

EDITED to add numbering on expected / actuals to simplify referencing the pieces of this issue.

Hi @breakwaterlabs!

I don't know what to say here. I'm doing your cases with 23.0.5 (last version).

In case 1 my result is the message: Error when trying to connect to LDAP: 'CommunicationError'.

![Screenshot from 2024-02-01 09-05-46](https://github.com/keycloak/keycloak/assets/26901232/d3e2a9f9-0c98-4ddc-8936-abea4644f6fa)

The wireshark if I change decode as TLS shows the hello:

![Screenshot from 2024-02-01 09-07-54](https://github.com/keycloak/keycloak/assets/26901232/b73d64ab-64c6-4a16-9b3d-a0d41c45de61)

And there is an exception showing the lines in which the test is performing a connection and a TLS handshake:

```
2024-02-01 09:03:32,553 DEBUG [org.keycloak.storage.ldap.idm.store.ldap.LDAPContextManager] (executor-thread-3) Creating LdapContext using properties: [{java.naming.factory.initial=com.sun.jndi.ldap.LdapCtxFactory, java.naming.provider.url=ldaps://1.1.1.1:80, com.sun.jndi.ldap.connect.timeout=20000, java.naming.security.authentication=none, java.naming.ldap.factory.socket=org.keycloak.truststore.SSLSocketFactory}]
2024-02-01 09:03:32,570 ERROR [org.keycloak.services] (executor-thread-3) KC-SERVICES0055: Error when connecting to LDAP: 1.1.1.1:80: javax.naming.CommunicationException: 1.1.1.1:80 [Root exception is javax.net.ssl.SSLException: Unsupported or unrecognized SSL message]
	at java.naming/com.sun.jndi.ldap.Connection.<init>(Connection.java:253)
	at java.naming/com.sun.jndi.ldap.LdapClient.<init>(LdapClient.java:137)
	at java.naming/com.sun.jndi.ldap.LdapClient.getInstance(LdapClient.java:1616)
	at java.naming/com.sun.jndi.ldap.LdapCtx.connect(LdapCtx.java:2848)
	at java.naming/com.sun.jndi.ldap.LdapCtx.<init>(LdapCtx.java:348)
	at java.naming/com.sun.jndi.ldap.LdapCtxFactory.getLdapCtxFromUrl(LdapCtxFactory.java:229)
	at java.naming/com.sun.jndi.ldap.LdapCtxFactory.getUsingURL(LdapCtxFactory.java:189)
	at java.naming/com.sun.jndi.ldap.LdapCtxFactory.getUsingURLs(LdapCtxFactory.java:247)
	at java.naming/com.sun.jndi.ldap.LdapCtxFactory.getLdapCtxInstance(LdapCtxFactory.java:154)
	at java.naming/com.sun.jndi.ldap.LdapCtxFactory.getInitialContext(LdapCtxFactory.java:84)
	at java.naming/javax.naming.spi.NamingManager.getInitialContext(NamingManager.java:732)
	at java.naming/javax.naming.InitialContext.getDefaultInitCtx(InitialContext.java:305)
	at java.naming/javax.naming.InitialContext.init(InitialContext.java:236)
	at java.naming/javax.naming.ldap.InitialLdapContext.<init>(InitialLdapContext.java:154)
	at org.keycloak.storage.ldap.idm.store.ldap.LDAPContextManager.createLdapContext(LDAPContextManager.java:80)
	at org.keycloak.storage.ldap.idm.store.ldap.LDAPContextManager.getLdapContext(LDAPContextManager.java:100)
	at org.keycloak.services.managers.LDAPServerCapabilitiesManager.testLDAP(LDAPServerCapabilitiesManager.java:139)
	at org.keycloak.services.resources.admin.TestLdapConnectionResource.testLDAPConnection(TestLdapConnectionResource.java:93)
	at org.keycloak.services.resources.admin.TestLdapConnectionResource$quarkusrestinvoker$testLDAPConnection_53600405a1d6ba020ad8536bd54994b827f01e74.invoke(Unknown Source)
	at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
	at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:141)
	at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:145)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:576)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: javax.net.ssl.SSLException: Unsupported or unrecognized SSL message
	at java.base/sun.security.ssl.SSLSocketInputRecord.handleUnknownRecord(SSLSocketInputRecord.java:457)
	at java.base/sun.security.ssl.SSLSocketInputRecord.decode(SSLSocketInputRecord.java:175)
	at java.base/sun.security.ssl.SSLTransport.decode(SSLTransport.java:111)
	at java.base/sun.security.ssl.SSLSocketImpl.decode(SSLSocketImpl.java:1510)
	at java.base/sun.security.ssl.SSLSocketImpl.readHandshakeRecord(SSLSocketImpl.java:1425)
	at java.base/sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:455)
	at java.base/sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:426)
	at java.naming/com.sun.jndi.ldap.Connection.createSocket(Connection.java:365)
	at java.naming/com.sun.jndi.ldap.Connection.<init>(Connection.java:232)
	... 28 more
```

Your case 3 is a bit misleading because StartTLS just makes sense with ldap (not ldaps). So your case 3 is doing exactly the same that case 1. It connects to a non TLS port, in my env it returns the same CommunicationError and it does not reach the point when the connection is established, so no sending the extended op (which probably would fail because it was already using TLS). So in my env this case 3 is exactly the same than case 1 (same images).

I don't know what's the difference, maybe the environments. But I'm also testing a direct connection against my localhost with an ldap server installed locally. I would suggest to test locally too with keycloak 23.0.5.

--- a/federation/ldap/src/main/java/org/keycloak/services/managers/LDAPServerCapabilitiesManager.java
+++ b/federation/ldap/src/main/java/org/keycloak/services/managers/LDAPServerCapabilitiesManager.java
@@ -18,6 +18,7 @@ package org.keycloak.services.managers;
 
 import java.util.Collections;
 import java.util.Set;
+import javax.naming.ldap.LdapContext;
 
 import org.jboss.logging.Logger;
 import org.keycloak.common.util.MultivaluedHashMap;
@@ -31,6 +32,7 @@ import org.keycloak.storage.ldap.LDAPConfig;
 import org.keycloak.representations.idm.LDAPCapabilityRepresentation;
 import org.keycloak.storage.ldap.idm.store.ldap.LDAPContextManager;
 import org.keycloak.storage.ldap.idm.store.ldap.LDAPIdentityStore;
+import org.keycloak.utils.StringUtil;
 
 /**
  * @author <a href="mailto:mposolda@redhat.com">Marek Posolda</a>
@@ -42,6 +44,21 @@ public class LDAPServerCapabilitiesManager {
     public static final String TEST_CONNECTION = "testConnection";
     public static final String TEST_AUTHENTICATION = "testAuthentication";
     public static final String QUERY_SERVER_CAPABILITIES = "queryServerCapabilities";
+    public static final int DEFAULT_TEST_TIMEOUT = 30000; // 30s default test timeout
+
+    private static int parseConnectionTimeout(String connectionTimeout) {
+        if (StringUtil.isNotBlank(connectionTimeout)) {
+            try {
+                int timeout = Integer.parseInt(connectionTimeout);
+                if (timeout > 0) {
+                    return timeout;
+                }
+            } catch (NumberFormatException e) {
+                // just use default timeout
+            }
+        }
+        return DEFAULT_TEST_TIMEOUT;
+    }
 
     public static LDAPConfig buildLDAPConfig(TestLdapConnectionRepresentation config, RealmModel realm) {
         String bindCredential = config.getBindCredential();
@@ -54,7 +71,11 @@ public class LDAPServerCapabilitiesManager {
         configMap.putSingle(LDAPConstants.BIND_CREDENTIAL, bindCredential);
         configMap.add(LDAPConstants.CONNECTION_URL, config.getConnectionUrl());
         configMap.add(LDAPConstants.USE_TRUSTSTORE_SPI, config.getUseTruststoreSpi());
-        configMap.putSingle(LDAPConstants.CONNECTION_TIMEOUT, config.getConnectionTimeout());
+        // set a forced timeout even when the timeout is infinite for testing
+        // this is needed to not wait forever in the test and force connection creation in ldap
+        String timeoutStr = Integer.toString(parseConnectionTimeout(config.getConnectionTimeout()));
+        configMap.putSingle(LDAPConstants.CONNECTION_TIMEOUT, timeoutStr);
+        configMap.putSingle(LDAPConstants.READ_TIMEOUT, timeoutStr);
         configMap.add(LDAPConstants.START_TLS, config.getStartTls());
         return new LDAPConfig(configMap);
     }
@@ -136,7 +157,11 @@ public class LDAPServerCapabilitiesManager {
         // Create ldapContextManager in try-with-resource so that ldapContext/tlsResponse/VaultSecret is closed/removed when it
         // is not needed anymore
         try (LDAPContextManager ldapContextManager = LDAPContextManager.create(session, ldapConfig)) {
-            ldapContextManager.getLdapContext();
+            LdapContext ldapContext = ldapContextManager.getLdapContext();
+            if (TEST_AUTHENTICATION.equals(config.getAction()) && LDAPConstants.AUTH_TYPE_NONE.equals(config.getAuthType())) {
+                // reconnect to force an anonymous bind operation
+                ldapContext.reconnect(null);
+            }
         } catch (Exception ne) {
             String errorMessage = (TEST_AUTHENTICATION.equals(config.getAction())) ? "Error when authenticating to LDAP: "
                 : "Error when connecting to LDAP: ";

--- a/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/UserFederationLdapConnectionTest.java
+++ b/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/UserFederationLdapConnectionTest.java
@@ -71,9 +71,9 @@ public class UserFederationLdapConnectionTest extends AbstractAdminTest {
         response = realm.testLDAPConnection(new TestLdapConnectionRepresentation(LDAPServerCapabilitiesManager.TEST_AUTHENTICATION, "ldap://localhost:10389", "uid=admin,ou=system", "${vault.ldap_bindCredential}", "false", null));
         assertStatus(response, 204);
 
-        // Authentication success anonymous bind
+        // Authentication error for anonymous bind (default ldap rule does not allow anonymous binings)
         response = realm.testLDAPConnection(new TestLdapConnectionRepresentation(LDAPServerCapabilitiesManager.TEST_AUTHENTICATION, "ldap://localhost:10389", null, null, "false", null, "false", LDAPConstants.AUTH_TYPE_NONE));
-        assertStatus(response, 204);
+        assertStatus(response, 400);
 
         response = realm.testLDAPConnection(new TestLdapConnectionRepresentation(LDAPServerCapabilitiesManager.TEST_AUTHENTICATION, "ldap://localhost:10389", "uid=admin,ou=system", "${vault.ldap_bindCredential}", "false", null));
         assertStatus(response, 204);
@@ -93,6 +93,15 @@ public class UserFederationLdapConnectionTest extends AbstractAdminTest {
         response = realm.testLDAPConnection(new TestLdapConnectionRepresentation(LDAPServerCapabilitiesManager.TEST_CONNECTION, "ldaps://localhostt:10636", "foo", "bar", "false", null));
         assertStatus(response, 400);
 
+        response = realm.testLDAPConnection(new TestLdapConnectionRepresentation(LDAPServerCapabilitiesManager.TEST_CONNECTION, "ldaps://localhost:10389", null, null, "false", null));
+        assertStatus(response, 400);
+
+        response = realm.testLDAPConnection(new TestLdapConnectionRepresentation(LDAPServerCapabilitiesManager.TEST_CONNECTION, "ldaps://localhost:10389", null, null, "false", "5000"));
+        assertStatus(response, 400);
+
+        response = realm.testLDAPConnection(new TestLdapConnectionRepresentation(LDAPServerCapabilitiesManager.TEST_CONNECTION, "ldaps://localhost:10389", null, null, "false", "0"));
+        assertStatus(response, 400);
+
         response = realm.testLDAPConnection(new TestLdapConnectionRepresentation(LDAPServerCapabilitiesManager.TEST_AUTHENTICATION, "ldaps://localhost:10636", "foo", "bar", "false", null));
         assertStatus(response, 400);
 
@@ -103,7 +112,7 @@ public class UserFederationLdapConnectionTest extends AbstractAdminTest {
         assertStatus(response, 204);
 
         response = realm.testLDAPConnection(new TestLdapConnectionRepresentation(LDAPServerCapabilitiesManager.TEST_AUTHENTICATION, "ldaps://localhost:10636", null, null, "false", null, "false", LDAPConstants.AUTH_TYPE_NONE));
-        assertStatus(response, 204);
+        assertStatus(response, 400);
 
         // Authentication success with bindCredential from Vault
         response = realm.testLDAPConnection(new TestLdapConnectionRepresentation(LDAPServerCapabilitiesManager.TEST_AUTHENTICATION, "ldaps://localhost:10636", "uid=admin,ou=system", "${vault.ldap_bindCredential}", "true", null));

